### PR TITLE
Adding OpenAI wrapper class to `notdiamond.toolkit.openai`. [ENG-1107]

### PR DESCRIPTION
**Pull request by acompa**
This PR creates sync and async OpenAI client wrappers for Not Diamond, allowing users to introduce ND-based routing to their code with one tiny change:

```python
# before
from openai import OpenAI

#after
from notdiamond.toolkit.openai import OpenAI
```

Tests are mostly designed to ensure functionality for the encapsulated OpenAI classes remains as expected.

### CODE REVIEW

**Comment by acompa** [comment]
Still working on async tests, please hold!

**Comment by acompa** [comment]
Async tests done, please feel free to review.

**Comment by r0ymanesco** [comment]
This this good as far as what it is meant to do. But I wonder if we might have set the wrong objective. If I look at what this user's [comment](https://not-diamond.slack.com/archives/C06N4HBPK9R/p1725847862754729) says, I think what they want to do is use the `OpenAI` client to do the same thing as the `NotDiamond` client. If this is the design objective, I wonder if all we need to do is use the `NotDiamond` client inside the `OpenAI` wrapper and simply map the `.create` method arguments to the `NotDiamond` client `.create` arguments. And either use a list or a `/` separator for specifying the list of model choices as the user's comment suggests. What do you think?

**Comment by acompa** [comment]
> I wonder if all we need to do is use the NotDiamond client inside the OpenAI wrapper and simply map the .create method arguments to the NotDiamond client .create arguments. And either use a list or a , separator for specifying the list of model choices as the user's comment suggests.

We cannot necessarily assume `.create` since that requires `notdiamond[create]` and therefore `langchain`. This new toolkit capability will be installed as `pip install notdiamond[openai]` instead; by omitting the `langchain` dependency there's a lower risk of conflicting with the user's runtime environment.

I didn't take the user's feedback as a hard requirement, since I personally don't love the `model="model1,model2,..."` approach they propose. But I agree that we should accept `model` as a keyword to `create`, so I can make that happen! The `openai.OpenAI` client doesn't use pydantic (as far as I know) to validate `model: str` anyway, so nothing keeps us from accepting multiple types for that kwarg value.

**Comment by r0ymanesco** [comment]
> > I wonder if all we need to do is use the NotDiamond client inside the OpenAI wrapper and simply map the .create method arguments to the NotDiamond client .create arguments. And either use a list or a , separator for specifying the list of model choices as the user's comment suggests.
> 
> We cannot necessarily assume `.create` since that requires `notdiamond[create]` and therefore `langchain`. This new toolkit capability will be installed as `pip install notdiamond[openai]` instead; by omitting the `langchain` dependency there's a lower risk of conflicting with the user's runtime environment. (Also, frankly, we don't need `langchain` to implement this when `openai` alone will suffice!)
> 
> I didn't take the user's feedback as a hard requirement, since I personally don't love the `model="model1,model2,..."` approach they propose. But I agree that we should accept `model` as a keyword to `create`, so I can make that happen! The `openai.OpenAI` client doesn't use pydantic (as far as I know) to validate `model: str` anyway, so nothing keeps us from accepting multiple types for that kwarg value.

Ok that's fair. We can make this client exclusively route between openai models and expand if we get user request. In that case I would say let's make the `model` argument accept a list or string`model: Union[str, List]` so that it can either be a single model or a list of models and map it to the `llm_providers` argument in ND client.

File: notdiamond/llms/request.py [hunk 1]
```diff
@@ -178,6 +178,9 @@ def model_select(
     )
 
     try:
+        print(url)
```
Comment (r0ymanesco) [inline notdiamond/llms/request.py:181]: I think you might have forgotten to remove these print statements

File: notdiamond/toolkit/openai.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        elif isinstance(model, str):
+            llm_configs = model.split(",")
+        elif isinstance(model, list):
+            llm_configs = self._nd_client._parse_llm_configs_data(model)
+
+        if "messages" not in kwargs:
+            raise ValueError("'messages' argument is required")
+
+        session_id, best_llm = self._nd_client.model_select(
+            *args, model=llm_configs, **kwargs
```
Comment (r0ymanesco) [inline notdiamond/toolkit/openai.py:101]: Shouldn't this be `llm_configs=llm_configs`?

File: notdiamond/toolkit/openai.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        elif isinstance(model, str):
+            llm_configs = model.split(",")
+        elif isinstance(model, list):
+            llm_configs = self._nd_client._parse_llm_configs_data(model)
+
+        if "messages" not in kwargs:
+            raise ValueError("'messages' argument is required")
+
+        session_id, best_llm = await self._nd_client.amodel_select(
+            *args, model=llm_configs, **kwargs
```
Comment (r0ymanesco) [inline notdiamond/toolkit/openai.py:195]: See above

File: notdiamond/toolkit/openai.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        )
+        LOGGER.info(f"Routed prompt to {best_llm} for session ID {session_id}")
```
Comment (r0ymanesco) [inline notdiamond/toolkit/openai.py:197]: Might want to do `LOGGER.debug` otherwise wouldn't this be printed everytime this method is called?

**Comment by r0ymanesco** [comment]
Tests seem to be failing but might have to do with the staging API key?

File: notdiamond/llms/request.py (hunk 1 above)
Comment (acompa) [reply notdiamond/llms/request.py:181]: Hard to write software while holding a sick baby, oy. Removed, thank you for catching.

File: notdiamond/toolkit/openai.py (hunk 4 above)
Comment (acompa) [reply notdiamond/toolkit/openai.py:197]: Yes, great point!

**Review by r0ymanesco — APPROVED** [review]
This looks good! I'm approving but noting that the tests are failing because of the ND api keys not working. Do these tests run against the staging instance or the prod instance?

**Comment by acompa** [comment]
TogetherAI test is failing, otherwise fine. Will merge this.

**Pull request merged**